Merge 2.12.x to 2.13.x [ci: last-only]

retronym opened the pull request:
```
$ git -P log $MB..origin/2.12.x --decorate --oneline --graph
```

// PICK
```
*   2ee5f6c783 (retronym/2.12.x, origin/2.12.x) Merge pull request #9500 from retronym/cleanup-cleanup
|\
| * f4da037f6b (retronym/cleanup-cleanup, origin/pr/9500) Fix non-exhaustive match in Cleanup phase
```

// SKIP (backport / 2.12.x specific)
```
*   eab3ce8693 Merge pull request #9507 from regadas/12344_bt_mima_report
|\
| * dc809a5649 (origin/pr/9507) Add missing MiMa filters for NewRedBlackTree
| * 0e0fb08da4 Fix mimaReportBinaryIssues, not run aggregate
|/
```

// PICK
```
* 01a8c458ed Merge pull request #9506 from retronym/topic/box-unbox-fix
* b705e3b467 (retronym/topic/box-unbox-fix, origin/pr/9506) Fix regression in box-unbox (in -opt:l:method)
```

// SKIP, backport
```
*   	 (origin/2.12.x) Merge pull request #9498 from bplommer/2.12.x
|\
| * 4a5d2b6dd2 Don't warnUnusedNowarn in scaladoc
```

// PICK
```
* |   83384b5086 Merge pull request #9499 from retronym/singleton-type-tree-pos
|\ \
| |/
|/|
| * a0f9840bfc (retronym/singleton-type-tree-pos, singleton-type-tree-pos) Retain positions of SingletonTypeTree
|/
```

// SKIP, already applied to 2.13
```
*   7bfcf50ede (origin/2.12.x) Merge pull request #9488 from SethTisue/asm-9.1
|\
| * d032ef8963 (origin/pr/9488) upgrade to asm 9.1, for JDK 17 support
|/
```

// PICK

```
*   f39a74abf0 Merge pull request #9475 from retronym/topic/stock-url-classloader
|\
| * 1e1a0fd6a9 (retronym/topic/stock-url-classloader, origin/pr/9475, topic/stock-url-classloader) Use a stock URLClassLoader in the scala runner for parallel capability
```

// SKIP, already applied 2.13.x
```
* |   3667278e85 Merge pull request #9477 from SethTisue/asm-9
|\ \
| * | 2a60c3121a (origin/pr/9477) upgrade to asm 9.0, for JDK 16 support
```

// PICK
```
* | |   adc9a22ef4 Merge pull request #9478 from retronym/ticket/12038-cyclic-found-reflection
|\ \ \
| |/ /
|/| |
| * | 845cda33e1 (retronym/ticket/12038-cyclic-found-reflection, origin/pr/9478, ticket/12038-cyclic-found-reflection) Fix cylic error in runtime reflection (protobuf)
|/ /
* |   5f6e5a1b5e (retronym/2.12.x, origin/pr/9469) Merge pull request #9461 from retronym/spire-array-regression
|\ \
| * | c9692130ae (retronym/spire-array-regression, origin/pr/9461, spire-array-regression) Faster `Array[T](t)` for @specialialized T.
| * | 3393616658 Fix regression in specialization of genericWrapArray(Array(..))
|/ /
```

// SKIP 2.12.x only
```
* |   a9b6d67e37 Merge pull request #9314 from mkeskells/2.12.x_RedBlackTree_serialisationAllocation
|\ \
| * | 646e6b7adb (origin/pr/9314) less allocation on deserialisation of RedBlackTree (SortedMap) reuse deserialisation logic for map and set
```


// PICK. 7d81f22f4e had non-trivial conflicts.
```
* | |   43a5a75bb1 Merge pull request #9082 from retronym/ticket/scala-async-152
|\ \ \
| * | | e309433964 (retronym/ticket/scala-async-152, origin/pr/9082, ticket/scala-async-152) async: await under synchronized is an error
* | | |   7ca86b97d5 Merge pull request #9455 from retronym/ticket/sd-741-pipeline-main-race
|\ \ \ \
| * | | | cf0b02796b (retronym/ticket/sd-741-pipeline-main-race, origin/pr/9455, ticket/sd-741-pipeline-main-race) Fix race condition in PipelineMainTest
| | |_|/
| |/| |
* | | |   f854b0f885 Merge pull request #9356 from retronym/ticket/11882
|\ \ \ \
| |/ / /
|/| | |
| * | | 7d81f22f4e (retronym/ticket/11882, origin/pr/9356, ticket/11882) Fix and extend Array.apply varargs optimization
```


// SKIP, 2.12.x specific
```
* | | |   fdb88812ac Merge pull request #9443 from som-snytt/issue/12315
|\ \ \ \
| * | | | 9a5fc06d52 (origin/pr/9443) Use Run reporter for wconf
|/ / / /
```


// PICK

```
* | | |   2a693ec702 Merge pull request #9355 from retronym/ticket/12213
|\ \ \ \
| |_|_|/
|/| | |
| * | | 50c2bb71c7 (retronym/ticket/12213, origin/pr/9355) Fix range position for vars with default initializer
| * | | 892a56d985 Test showing position bug with `var x = _`
| |/ /
```

// SKIP
```
* | |   2f1a80763d Merge pull request #9436 from dwijnand/2.12/mima
|\ \ \
| * | | 789fe0dc93 (origin/pr/9436) [bport] Move & truncate the MiMa filters
```


// PICK
```
* | | |   e72a873f99 Merge pull request #9315 from mkeskells/2.12.x_RedBlack_reuse_fp
|\ \ \ \
| * | | | 0678aeaa9a (origin/pr/9315) reuse common code for filter/partition for RedBlackTree based Set/Map
|/ / / /
* | | |   951c395c80 Merge pull request #9316 from mkeskells/2.12.x_RedBlack_simple
|\ \ \ \
| * | | | b990bff0e9 (origin/pr/9316) Deletion from a TreeSet/TreeMap of a key that doesn't exist should not allocate pass the tree into balance/balLeft/balRight methods so that is can be reused is unchanged
* | | | |   d613e1f5f9 Merge pull request #9046 from mkeskells/2.12.x_RedBlackTree_remove_duplication
|\ \ \ \ \
| |_|/ / /
|/| | | |
| * | | | bc742216fd (origin/pr/9046) de-duplicate code code filter and partition is common other than the supplied function
| |/ / /
```

// PICK
```
* | | |   78bf3a2d1e Merge pull request #9342 from gzm0/annotated-literal-annotation-arg
|\ \ \ \
| |/ / /
|/| | |
| * | | 2297931588 (origin/pr/9342) Allow annotated literals inside as annotation args
```

// PICK
```
* | | |   85a8814d26 Merge pull request #9369 from sjrd/remove-dead-code-patmat-array-unit
|\ \ \ \
| * | | | f39222d7bf (origin/pr/9369) Remove dead code testing for Array[Unit] after Array[AnyRef].
```

// PICK 

```
* | | | |   d057f62275 Merge pull request #9419 from koterpillar/patch-1
|\ \ \ \ \
| * | | | | a398b9b939 (origin/pr/9419) Enumeration documentation: remove nested class shadowing in "override" position
| | |_|_|/
| |/| | |
```

// PICK 
```
* | | | |   f2d2e6e6e0 Merge pull request #9426 from tanishiking/singleton-type-tree-12296
|\ \ \ \ \
| * | | | | 8a460167cf (origin/pr/9426) Ensure SingletonTypeTree#ref has symbol after typing.
| |/ / / /
* | | | |   8850b1fe0c Merge pull request #9434 from aaabramov/patch-1
|\ \ \ \ \
| * | | | | e5ff46e361 (origin/pr/9434) Fix typo in comment in Names.scala
```

// PICK

```
* | | | | |   ff03e5ad95 Merge pull request #9433 from lrytz/box-unbox-branches
|\ \ \ \ \ \
| |/ / / / /
|/| | | | |
| * | | | | 2ebf54dafe (origin/pr/9433) Fix box-unbox around branches
|/ / / / /
```


// SKIP backports and 2.12.13 release commits
```
* | | | |   09c9ef257d Merge pull request #9393 from lrytz/nowarn-nowarn
|\ \ \ \ \
| * | | | | e5580a6f61 (origin/pr/9393) [nomerge] silence warning on nowarn
* | | | | |   dd0a576ced Merge pull request #9432 from SethTisue/post-2.12.13
|\ \ \ \ \ \
| * | | | | | 457c7b44c4 (origin/pr/9432) re-STARR on 2.12.13
* | | | | | |   fd08bb447f Merge pull request #9420 from SethTisue/copyright-2021
|\ \ \ \ \ \ \
| |/ / / / / /
|/| | | | | |
| * | | | | | 906f4cd178 (origin/pr/9420) bump copyright year to 2021
| | |/ / / /
| |/| | | |
* | | | | | 60b503bcfd (tag: v2.12.13) give sonatype more time to close staging repos
* | | | | |   dfc691a5f3 Merge pull request #9430 from lrytz/build-fixes
|\ \ \ \ \ \
| |/ / / / /
|/| | | | |
| * | | | | 288926661d (origin/pr/9430) Fix release script for sbt 1.x update
|/ / / / /
```


// SKIP not applicable to groupBy in 2.13.x, the previous optimization is not yet forward ported.
```
* | | | |   b211dc4e90 Merge pull request #9376 from mkeskells/2.12.x_optimise_optimised_groupBy
|\ \ \ \ \
| * | | | | 5f0210567f (origin/pr/9376) remove some tuples
| | |_|/ /
| |/| | |
```

// SKIP backport

```
* | | | |   f5f5aae91a Merge pull request #9413 from SethTisue/travis-ci-com-2.12
|\ \ \ \ \
| * | | | | 394e6d4d0b (origin/pr/9413) [backport] move from travis-ci.org to travis-ci.com
|/ / / / /
```

// SKIP merge and immediate revert
```
* | | | |   2d69618f5e Merge pull request #9403 from lrytz/revert9392
|\ \ \ \ \
| * | | | | 4d30d48d7f (origin/pr/9403) Revert "Fix box-unbox around branches"
|/ / / / /
* | | | |   bb5e8c248b Merge pull request #9392 from lrytz/box-unbox-branches
|\ \ \ \ \
| |_|/ / /
|/| | | |
| * | | | 13a4254830 (origin/pr/9392) Fix box-unbox around branches
|/ / / /
```

// SKIP f4f6491306 was included in the forward port of that feature already. Others are backports.

```
* | | |   a90564e3ff Merge pull request #9382 from SethTisue/moar-heap
|\ \ \ \
| * | | | bd800153a3 (origin/pr/9382) [backport] backport .jvmopts from 2.13.x, sbt needs heap
|/ / / /
* | | |   471ea9d5a5 Merge pull request #9371 from retronym/2.12.x-partest-publish
|\ \ \ \
| |_|/ /
|/| | |
| * | | 58c58212c4 (retronym/2.12.x-partest-publish, origin/pr/9371) Remove non existing dependencies from scala-partest
| * | | cb865d353d Build scaladocs for partest to make sonatype happy
|/ / /
* | |   8b13dab7c8 Merge pull request #9368 from lrytz/backport-travis-wait
|\ \ \
| * | | 96db33a882 (origin/pr/9368) [backport] Avoid silence during CI, which Travis finds unacceptable
|/ / /
* | |   5c9699d081 Merge pull request #9361 from retronym/faster/class-tag-cache-property-check
|\ \ \
| * | | f4f6491306 (retronym/faster/class-tag-cache-property-check, origin/pr/9361, faster/class-tag-cache-property-check) Only look up the system prop controlling class tag caching once
| | |/
| |/|
* | |   0a3bb243ea Merge pull request #9358 from retronym/ticket/remove-pack200
|\ \ \
| |/ /
|/| |
| * | 26ceba64ac (retronym/ticket/remove-pack200, origin/pr/9358, ticket/remove-pack200) Remove pack200 ant task to make 2.12.x compilable on JDK 15
|/ /
```

// PICK

```
* |   ade1bbc42e Merge pull request #9325 from sadhen/bug_11247
|\ \
| |/
|/|
| * 398a94c235 (origin/pr/9325) Logging
| * dd5782f8e0 Catch exceptions
```


// SKIP, all backports of bug fixes in non 2.13.x code.

```
* |   aa73e74465 (mkeskells/2.12.x) Merge pull request #9335 from SethTisue/fix-array-empty-unit
|\ \
| * | 5e6676d996 (origin/pr/9335) fix regression where Array[Unit].empty caused NPE
|/ /
* |   5ee239ef63 Merge pull request #9321 from SethTisue/deprecate-arraycharsequence
|\ \
| * | b10abb48eb (origin/pr/9321) deprecate ArrayCharSequence
|/ /
* |   44be245a4a Merge pull request #9322 from SethTisue/backport-9292
|\ \
| |/
|/|
| * 6d609f1862 (origin/pr/9322) [backport] Make the CharSequence wrappers in Predef non-implicit, for JDK 15
|/
*   30c5b82ab4 Merge pull request #9311 from retronym/topic/sorted-map-set-bincompat
|\
| * b5e6d54a28 (retronym/topic/sorted-map-set-bincompat, origin/pr/9311) Avoid use of binary incompatbile super accessors in SortedMap/Set
|/
*   2166efcd2b Merge pull request #9276 from jtjeferreira/class-tag-leak
|\
| * c18fb61ac7 (origin/pr/9276) Avoid possible NPE in ClassTag cache's weak reference
| * 7fb5a1b8ce fix ClassTag cache leak
* 76f197e20c Merge pull request #9310 from retronym/backport/extra-lazy-annotation-info
* 79ae260359 (retronym/backport/extra-lazy-annotation-info, origin/pr/9310) [backport] Make annotation typechecking lazier
```

lrytz committed:
Build scaladocs for partest to make sonatype happy

(cherry picked from commit a0026305125dcc814097b1fbb798fe925ca9f81f)

lrytz committed:
Remove non existing dependencies from scala-partest

The `scala-repl-jline-embedded` and `scala-compiler-doc` sbt modules are
merged into `scala-compiler`.

(cherry picked from commit 9437be7cc8bb3e63d66af31fb22d62983acf0c22)

retronym committed:
[backport] Avoid silence during CI, which Travis finds unacceptable

lrytz committed:
Fix box-unbox around branches

  - Support `pop` as BoxConsumer
  - consumer of box creation may be the consumer for multiple values

Inspired by the tuple cases; extended to the ref/primitive cases.

Co-Authored-By: "Harrison Houghton" <hora.rhino@gmail.com>

lrytz committed:
Fix box-unbox around branches

  - Support `pop` as BoxConsumer
  - consumer of box creation may be the consumer for multiple values

Inspired by the tuple cases; extended to the ref/primitive cases.

Co-Authored-By: "Harrison Houghton" <hora.rhino@gmail.com>

lrytz committed:
Don't warnUnusedNowarn in scaladoc

mkeskells committed:
de-duplicate code code
filter and partition is common other than the supplied function

retronym committed:
async: await under synchronized is an error

Co-authored-by: Lukas Rytz <lukas.rytz@gmail.com>

jtjeferreira committed:
fix ClassTag cache leak

JDK issues about the problem:
* https://bugs.openjdk.java.net/browse/JDK-8136353

dwijnand committed:
[backport] Make the CharSequence wrappers in Predef non-implicit, for JDK 15

In JDK 15 CharSequence has an isEmpty method with a default
implementation, which clashes with our Array[Char]#isEmpty,
IndexedSeq[Char]#isEmpty, as well as our StringBuilder and reflect's
Name.

Backport of #9292 from 2.13.x to 2.12.x

retronym committed:
Avoid possible NPE in ClassTag cache's weak reference

retronym committed:
[backport] Make annotation typechecking lazier

Now that `Typer.stabilize` ends up checking for `@uncheckedStable`
annotations to support defs's as stable paths, new opportunities
arise for cyclic errors.

This PR reduces the likelihood of cycles by typechecking the only core of
an annotation (_not_ the full application to the annotation arguments)
to determine its type symbol.

pos/unchecked-stable-cyclic-error.scala test case started to fail
since #8338 with:

```
|-- object Test BYVALmode-EXPRmode (site: package <empty>)
|    |-- super APPSELmode-EXPRmode-POLYmode-QUALmode (silent: <init> in Test)
|    |    |-- this EXPRmode (silent: <init> in Test)
|    |    |    \-> Test.type
|    |    \-> Test.type
|    |-- def foo BYVALmode-EXPRmode (site: object Test)
|    |    |-- attr EXPRmode (site: method foo in Test)
|    |    |    |-- Int TYPEmode (site: method attr in Test)
|    |    |    |    \-> Int
|    |    |    |-- new anno APPSELmode-BYVALmode-EXPRmode-FUNmode-POLYmode (site: object Test)
|    |    |    |    |-- new anno APPSELmode-EXPRmode-POLYmode-QUALmode (site: object Test)
|    |    |    |    |    |-- anno FUNmode-TYPEmode (site: object Test)
|    |    |    |    |    |    \-> anno
|    |    |    |    |    \-> anno
|    |    |    |    \-> (a: Any): anno
|    |    |    |-- (a: Any): anno : pt=anno EXPRmode (site: value <local Test> in Test)
|    |    |    |    |-- foo : pt=Any BYVALmode-EXPRmode (site: value <local Test> in Test)
|    |    |    |    |    caught scala.reflect.internal.Symbols$CyclicReference: illegal cyclic reference involving method foo: while typing foo
/Users/jz/code/scala/sandbox/test.scala:7: error: recursive method foo needs result type
  @anno(foo) def attr: Int = foo
        ^
|    |    |    |    \-> anno
```

Another variant, pos/annotation-cycle.scala, fails only on 2.12.x
(after the backport of #8338). Backporting this fix makes that test
start passing on 2.12.x.

Backports #9302

retronym committed:
Merge pull request #9310 from retronym/backport/extra-lazy-annotation-info

[backport] Make annotation typechecking lazier

retronym committed:
Avoid use of binary incompatbile super accessors in SortedMap/Set

retronym committed:
Merge pull request #9276 from jtjeferreira/class-tag-leak

fix ClassTag#cache leak

mkeskells committed:
less allocation on deserialisation of RedBlackTree (SortedMap)
reuse deserialisation logic for map and set

mkeskells committed:
reuse common code for filter/partition for RedBlackTree based Set/Map

retronym committed:
Merge pull request #9311 from retronym/topic/sorted-map-set-bincompat

Avoid use of binary incompatible super accessors in SortedMap/Set

mkeskells committed:
Deletion from a TreeSet/TreeMap of a key that doesn't exist should not allocate
pass the tree into balance/balLeft/balRight methods so that is can be reused is unchanged

prefer use of isBlack/isRed to isRedTree when we know the tree is not null

lrytz committed:
Merge pull request #9322 from SethTisue/backport-9292

[backport] Make the CharSequence wrappers in Predef non-implicit, for JDK 15

da-liii committed:
Catch exceptions

SethTisue committed:
deprecate ArrayCharSequence

as per discussion on scala/scala#9292

SethTisue committed:
Merge pull request #9321 from SethTisue/deprecate-arraycharsequence

deprecate ArrayCharSequence

da-liii committed:
Logging

dwijnand committed:
[bport] Move & truncate the MiMa filters

Backport cherry-pick of 710c91e076e7c4984b011e9777bfa5f45cbb401d.

First of all, move the damn MiMa filters outside of build.sbt!  In
build.sbt it overpowers the rest of the build definition and also (it
would seem) IntelliJ is better at handling Scala files than sbt files.

Now, if a contributor needs to add filters, they can be advised to
change "project/MimaFilters.scala".

Secondly, bump the `mimaReferenceVersion` (which is used to define
`mimaPreviousArtifacts`).  Normal projects (without forwards
compatibility) want to always bump that, so that new methods don't go
missing.  But for Scala it's less of a problem.  However, it is because
it means we have an always appended list of exclusions, and those might
end up hiding real problems.

While I was at it I fixed `mimaReportBinaryIssues` it can be called
directly, by resetting `ThisBuild / mimaFailOnNoPrevious`.

SethTisue committed:
fix regression where Array[Unit].empty caused NPE

this regressed in #9091

the problem turned up in the Scala 2.12 community build -- when Scala
3 support was added to the scala-collection-compat repo, a test case
there was altered in a way that happened to tickle this

SethTisue committed:
Merge pull request #9335 from SethTisue/fix-array-empty-unit

fix never-released 2.12.13 regression where Array[Unit].empty caused NPE

gzm0 committed:
Allow annotated literals inside as annotation args

This notably allows for silencing warnings inside annotation
arguments (see scala-js/scala-js#3737, ghik/silencer#32).

Further, this fixes scala/bug#11932.

retronym committed:
Merge pull request #9325 from sadhen/bug_11247

Catch exceptions for the InlineInfoAttribute reader

retronym committed:
Test showing position bug with `var x = _`

The `= _` should be include in the range position
of the ValDef.

retronym committed:
Fix range position for vars with default initializer

Pass the position of the RHS explicitly rather than
using `rhs.pos` within TreeBuilder, as that doesn't
carry the position of the `_` represented by the
singleton tree `EmptyTree`.

Change the quasiquote parser to also pick up the change.

Also, tweaks in DefinitionConstructionProps that came out of fixing the
quasiquote parser change.

Co-authored-by: Dale Wijnand <dale.wijnand@gmail.com>

retronym committed:
Fix and extend Array.apply varargs optimization

The existing optimization for `Array.apply[T]` overreaches by
assuming that the `ArrayValue` carrying the varargs will have the
same element type as the implicit class tag. However, this is
not the case when the type argument is a generic type:

```
scala> def foo[T <: AnyRef : reflect.ClassTag](t: T) = Array.apply[T](t); foo[String]("")
java.lang.ClassCastException: [Ljava.lang.Object; cannot be cast to [Ljava.lang.String;
  ... 32 elided
```

This PR limits the existing optimization to cases where:

  - The class tag is an implicit arg
  - The class tag is materialized (not found via normal implicit search)

The second criterion excludes the call above.

A new, fallback optimization is provided that avoids a temporary
array and collection wrapper by emitting:

```
val arr = classTag.newArray(N)
ScalaRunTime.arrayUpdate(arr, N, elemN)
...
arr
}
```

retronym committed:
Remove pack200 ant task to make 2.12.x compilable on JDK 15

Removed from the JDK: https://openjdk.java.net/jeps/367

Ant support has been removed totally in 2.13.x, but remains in
2.12.x. Of the small number of remaining 2.12.x / Ant users, I suspect
the number using the pack200 task is approximately zero.

retronym committed:
Only look up the system prop controlling class tag caching once

dwijnand committed:
Merge pull request #9358 from retronym/ticket/remove-pack200

Remove pack200 ant task to make 2.12.x compilable on JDK 15

SethTisue committed:
Merge pull request #9361 from retronym/faster/class-tag-cache-property-check

Only look up the system prop controlling class tag caching once

lrytz committed:
Merge pull request #9368 from lrytz/backport-travis-wait

[backport] Avoid silence during CI, which Travis finds unacceptable

sjrd committed:
Remove dead code testing for Array[Unit] after Array[AnyRef].

Since `Array[Unit]` is implemented as `Array[BoxedUnit]`, any
`Array[Unit]` already qualifies as an `Array[AnyRef]`. Therefore,
all the removed cases were dead code.

lrytz committed:
Merge pull request #9371 from retronym/2.12.x-partest-publish

mkeskells committed:
remove some tuples

SethTisue committed:
[backport] backport .jvmopts from 2.13.x, sbt needs heap

SethTisue committed:
Merge pull request #9382 from SethTisue/moar-heap

[backport] backport .jvmopts from 2.13.x, sbt needs heap

lrytz committed:
[nomerge] silence warning on nowarn

lrytz committed:
Merge pull request #9392 from lrytz/box-unbox-branches

Fix box-unbox around branches

lrytz committed:
Revert "Fix box-unbox around branches"

This reverts commit 13a425483099f390e8ef4a939736480fce9a14a4, to be
re-applied after 2.12.13 is released.

lrytz committed:
Merge pull request #9403 from lrytz/revert9392

Revert "Fix box-unbox around branches"

SethTisue committed:
[backport] move from travis-ci.org to travis-ci.com

backport of #9411 to 2.12.x
also backports 98acd0f

context: scala/scala-dev#728

SethTisue committed:
Merge pull request #9413 from SethTisue/travis-ci-com-2.12

[backport] move from travis-ci.org to travis-ci.com

diesalbla committed:
Merge pull request #9376 from mkeskells/2.12.x_optimise_optimised_groupBy

optimize the optimized groupBy

koterpillar committed:
Enumeration documentation: remove nested class shadowing in "override" position

SethTisue committed:
bump copyright year to 2021

in the usual places. based on 6de01bc, and verified with
`git grep -w 2020`

tanishiking committed:
Ensure SingletonTypeTree#ref has symbol after typing.

Fixes scala/bug#12296

lrytz committed:
Fix release script for sbt 1.x update

Backport of scala/scala#7274

lrytz committed:
Merge pull request #9430 from lrytz/build-fixes

Fix release script for sbt 1.x update

lrytz committed:
give sonatype more time to close staging repos

SethTisue committed:
Merge pull request #9420 from SethTisue/copyright-2021

bump copyright year to 2021

SethTisue committed:
re-STARR on 2.12.13

lrytz committed:
Merge pull request #9432 from SethTisue/post-2.12.13

re-STARR on 2.12.13

lrytz committed:
Merge pull request #9393 from lrytz/nowarn-nowarn

[nomerge] silence warning on nowarn

aaabramov committed:
Fix typo in comment in Names.scala

lrytz committed:
Merge pull request #9433 from lrytz/box-unbox-branches

Fix box-unbox around branches

SethTisue committed:
Merge pull request #9434 from aaabramov/patch-1

Fix typo in comment in Names.scala

SethTisue committed:
Merge pull request #9426 from tanishiking/singleton-type-tree-12296

dwijnand committed:
Merge pull request #9419 from koterpillar/patch-1

Enumeration documentation: remove nested class shadowing in "override" position

dwijnand committed:
Merge pull request #9369 from sjrd/remove-dead-code-patmat-array-unit

Remove dead code testing for Array[Unit] after Array[AnyRef].

dwijnand committed:
Merge pull request #9342 from gzm0/annotated-literal-annotation-arg

Allow annotated literals inside as annotation args

dwijnand committed:
Merge pull request #9046 from mkeskells/2.12.x_RedBlackTree_remove_duplication

Remove duplicated code in RedBlackTree

SethTisue committed:
Merge pull request #9316 from mkeskells/2.12.x_RedBlack_simple

remove some allocations from RedBlackTree

dwijnand committed:
Merge pull request #9315 from mkeskells/2.12.x_RedBlack_reuse_fp

reuse common code for filter/partition for RedBlackTree based Set/Map

dwijnand committed:
Merge pull request #9436 from dwijnand/2.12/mima

[bport] Move & truncate the MiMa filters

dwijnand committed:
Merge pull request #9355 from retronym/ticket/12213

som-snytt committed:
Use Run reporter for wconf

lrytz committed:
Merge pull request #9443 from som-snytt/issue/12315

Use Run reporter for wconf

retronym committed:
Fix race condition in PipelineMainTest

This stack, for example, touches the shared reporter from a background
thread, so we need to add synchronization.

```
  scala.reflect.internal.Reporter.echo(Reporting.scala:109)
  scala.tools.nsc.PipelineMainClass.scala$tools$nsc$PipelineMainClass$$reporterEcho(PipelineMain.scala:63)
  scala.tools.nsc.PipelineMainClass$Task.log(PipelineMain.scala:586)
  scala.tools.nsc.PipelineMainClass$Task.javaCompile(PipelineMain.scala:548)
  scala.tools.nsc.PipelineMainClass.$anonfun$process$40(PipelineMain.scala:214)
  scala.tools.nsc.PipelineMainClass.$anonfun$process$40$adapted(PipelineMain.scala:212)
  scala.util.Success.$anonfun$map$1(Try.scala:255)
  scala.util.Try$.apply(Try.scala:213)
  scala.util.Success.map(Try.scala:255)
  scala.concurrent.Future.$anonfun$map$1(Future.scala:292)
  scala.concurrent.impl.Promise.liftedTree1$1(Promise.scala:33)
  scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:33)
  scala.concurrent.impl.CallbackRunnable.run(Promise.scala:64)
  java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1402)
```

retronym committed:
Merge pull request #9356 from retronym/ticket/11882

Fix and extend Array.apply varargs optimization

lrytz committed:
Merge pull request #9455 from retronym/ticket/sd-741-pipeline-main-race

Fix race condition in PipelineMainTest

lrytz committed:
Merge pull request #9082 from retronym/ticket/scala-async-152

async: await under synchronized is an error

SethTisue committed:
Merge pull request #9314 from mkeskells/2.12.x_RedBlackTree_serialisationAllocation

less allocation on deserialisation of RedBlackTree (SortedMap)

dwijnand committed:
Fix mimaReportBinaryIssues, not run aggregate

When you run a task in the shell, it uses the current project (normally
root) and runs the tasks on the project and the project it aggregates.
That's why "compile" compiles all and "test" tests all.  But "testAll"
is our custom thing, defined only on the root project, which refers to
specific tasks to run.  So while just "mimaReportBinaryIssues" does the
right thing in the shell, it only checks root (which is a noop) when
used in "testAll".  So back to specifying both - but at least I still
fixed unqualified "mimaReportBinaryIssues" from working in the shell.

(cherry picked from commit 8e5951d307bfef14cb1a19c50196dd55b1ff4441)
backport of #9459

retronym committed:
Fix regression in specialization of genericWrapArray(Array(..))

The extension to the optimization in Cleanup of Array.apply
assumed that the argument array was non-primitive.  This broke
compilation of spire (code minimized to test/files/run/sd760a.scala)

I also found a path to the same bug without specialization, by using
a generic type parmater bounded by a primitive. This is in
test/files/run/sd760b.scala

The fix here is straight forward -- disable this optimization when the
argument array is primitive.

retronym committed:
Faster `Array[T](t)` for @specialialized T.

Avoid boxing the elements in favour of bulk clone of the varargs
array into the result array, when the runtime type of the
class-tag-generated result array is the same as the varargs array.

SethTisue committed:
Merge pull request #9461 from retronym/spire-array-regression

retronym committed:
Use a stock URLClassLoader in the scala runner for parallel capability

Background: Java 7 introduced a way for classloaders to use
fine grained locking to avoid some deadlocks possibilities.
The same fine grained locks could conceivably improve performance
of multi-threaded code that has non-trivial logic in static
initializers.

This tech note describes the change:
  https://docs.oracle.com/javase/7/docs/technotes/guides/lang/cl-mt.html

For backwards compatibility, custom subclasses of ClassLoader need to
opt into this scheme by calling `ClassLoader.registerParalleCapable`.
This needs to be done from the class itself during it static
initializer (well, at least before the first super constructor call to
`ClassLoader(...)`)

The Scala runners use a custom classloader subclass. This has never
opted into the parallel classloading.

It is extremely difficult to make ScalaClassLoader.URLClassLoader
correctly register itself to ClassLoader.parallelCapable due to
the absence of Scala language support for declaring Java static
initializer blocks. Implementing tbis class in Java is possible
but in turn requires our SBT build to no longer use the
"JavaThenScala" compilation mode.

However, closer inspection reveals that this custom classloader type is
only used for convenience methods.

This PR refactors these into extension methods and uses a plain
`java.net.URLClassLoader` directly.

The changes is observable with:

```scala
object Test extends App {
  val cl = Test.getClass.getClassLoader
  val fld = classOf[java.lang.ClassLoader].getDeclaredField("parallelLockMap")
  fld.setAccessible(true)
  val isParallel = fld.get(cl) != null
  println((cl, "isParallel = " + isParallel))
}
```

```
% (export V=2.12.12; scalac --scala-version $V  -d /tmp sandbox/test.scala && scala --scala-version $V -cp /tmp Test )
(scala.reflect.internal.util.ScalaClassLoader$URLClassLoader@58c1c010,isParallel = false)
% qscalac -d /tmp sandbox/test.scala && qscala -cp /tmp Test
(java.net.URLClassLoader@cac736f,isParallel = true)
```

retronym committed:
Fix cylic error in runtime reflection (protobuf)

The use of f-bounded generic in protobuf that refer to
static inner classes started causing cyclic errors in
runtime reflecion since scala/scala#12038.

We can avoid the problem by limiting the use of the
`ParameterizedType#getOwnerType` to cases when the selected inner
class is non-static.

SethTisue committed:
upgrade to asm 9.0, for JDK 16 support

fixes scala/bug#12328

lrytz committed:
Merge pull request #9478 from retronym/ticket/12038-cyclic-found-reflection

Fix cylic error in runtime reflection (protobuf)

lrytz committed:
Merge pull request #9477 from SethTisue/asm-9

upgrade to asm 9.0, for JDK 16 support

lrytz committed:
Merge pull request #9475 from retronym/topic/stock-url-classloader

Use a stock URLClassLoader in the scala runner for parallel capability

SethTisue committed:
upgrade to asm 9.1, for JDK 17 support

fixes scala/bug#12332

SethTisue committed:
Merge pull request #9488 from SethTisue/asm-9.1

upgrade to asm 9.1, for JDK 17 support

retronym committed:
Merge commit 'aa73e74465' into merge-2.12.x-to-2.13.x-20210211

retronym committed:
Merge commit 'ade1bbc42e' into merge-2.12.x-to-2.13.x-20210211

retronym committed:
Merge commit 'f5f5aae91a' into merge-2.12.x-to-2.13.x-20210211

retronym committed:
Merge commit '09c9ef257d' into merge-2.12.x-to-2.13.x-20210211

retronym committed:
Merge commit 'ff03e5ad95' into merge-2.12.x-to-2.13.x-20210211

retronym committed:
Merge commit 'f2d2e6e6e0' into merge-2.12.x-to-2.13.x-20210211

retronym committed:
Merge commit 'd057f62275' into merge-2.12.x-to-2.13.x-20210211

retronym committed:
Merge commit '85a8814d26' into merge-2.12.x-to-2.13.x-20210211

retronym committed:
Merge commit '78bf3a2d1e' into merge-2.12.x-to-2.13.x-20210211

retronym committed:
Merge commit 'e72a873f99' into merge-2.12.x-to-2.13.x-20210211

retronym committed:
Merge commit '2f1a80763d' into merge-2.12.x-to-2.13.x-20210211

retronym committed:
Merge commit '2a693ec702' into merge-2.12.x-to-2.13.x-20210211

And Fixup RedBlackTree merge

SethTisue commented:
@retronym as per https://issues.apache.org/jira/browse/SPARK-25075, the Spark folks consider https://github.com/scala/bug/issues/12038 to be a blocker for their 2.13 upgrade, so it would be good to be extra-double-sure that this merge gets https://github.com/scala/scala/pull/9478 into 2.13.5

retronym committed:
Retain positions of SingletonTypeTree

Regressed in #9426. I noticed the missing position when merging
that change forward to 2.13.x, in which a new test case
(neg/t12322a.scala) started giving inaccurate positions.

lrytz committed:
Merge pull request #9499 from retronym/singleton-type-tree-pos

Retain positions of SingletonTypeTree

dwijnand committed:
Merge pull request #9498 from bplommer/2.12.x

Don't warnUnusedNowarn in scaladoc (scala 2.12)

retronym committed:
Merge commit 'fdb88812ac' into merge-2.12.x-to-2.13.x-20210211

retronym committed:
Merge commit '43a5a75bb1' into merge-2.12.x-to-2.13.x-20210211

retronym committed:
Merge commit 'a9b6d67e37' into merge-2.12.x-to-2.13.x-20210211

retronym committed:
Merge commit 'adc9a22ef4' into merge-2.12.x-to-2.13.x-20210211

retronym committed:
Fixup async synchronized checkfile

retronym committed:
Merge commit '3667278e85' into merge-2.12.x-to-2.13.x-20210211

retronym committed:
Merge commit 'f39a74abf0' into merge-2.12.x-to-2.13.x-20210211

retronym committed:
Merge commit '7bfcf50ede' into merge-2.12.x-to-2.13.x-20210211

retronym committed:
Merge commit '83384b5086' into merge-2.12.x-to-2.13.x-20210211

retronym committed:
Merge commit '8562bab041' into merge-2.12.x-to-2.13.x-20210211

retronym committed:
Fix non-exhaustive match in Cleanup phase

(cherry picked from commit 038b26de049c7e35d52dc66e773ca114cb5434e8)

retronym commented:
The bootstrap failure minimizes to:

```scala
package p1

class Test {

  def crash: Unit = {
    var box: AnyRef = null

    @inline def writeBox(a: AnyRef): Unit = {
      box = a
    }

    writeBox(null)
  }
}
```

```
% qscalac -d /tmp -J-Xmx1G -Yopt-trace 'p1/Test.crash' -opt:l:inline -opt-inline-from:'**/*' /tmp/HashMap.scala

/tmp/HashMap.scala:3: error: Error while emitting p1/Test
scala.tools.asm.tree.analysis.AnalyzerException: While processing p1/Test.crash
...
Caused by: scala.tools.asm.tree.analysis.AnalyzerException: Error at instruction 20: Cannot pop operand off an empty stack.
	at scala.tools.asm.tree.analysis.Analyzer.analyze(Analyzer.java:296)
	at scala.tools.nsc.backend.jvm.analysis.AsmAnalyzer.<init>(AsmAnalyzer.scala:28)
	... 54 more
Caused by: java.lang.IndexOutOfBoundsException: Cannot pop operand off an empty stack.
	at scala.tools.asm.tree.analysis.Frame.pop(Frame.java:258)
```

Here's the bytecode diff introduced by `BoxUnbox` in 2.13.x. Notice that an `ASTORE` is dropped on the grounds that is a (transitive) copy operation of the box that was being eliminated. But the following `POP` was not removed.

https://gist.github.com/retronym/d4091c9ce792cb2bab24b29e8312032c/revisions#diff-6eb5aa2af122281945d8d42cc285a04af319c927fec0b2e03ce47987c39985a2L25

This is related to the change in the backend around `BoxUnbox`. This doesn't crash on 2.12.x, so it could be a merging error. But it could also be a bug in 2.12.x that is hidden for some other reason.

/cc @lrytz

retronym commented:
Same crash, slightly simpler bytecode. 

```scala
package p1

class Test {

  def crash: Unit = {
    var box = 0

    @inline def writeBox: Unit = {
      box = 1
    }

    writeBox
  }
}
```

https://gist.github.com/retronym/2e285db92cc4f967d68c70b532010c80/revisions#diff-6eb5aa2af122281945d8d42cc285a04af319c927fec0b2e03ce47987c39985a2L26

I don't think the inliner is essential, but so far I haven't found a repro without it.

retronym commented:
I've pushed a commit with a candidate fix. I imagine this would also be needed on 2.12.x but maybe we never hit that code path in either the 2.12.x test suite of bootstrap?

retronym commented:
There is a second crash with:

```scala
package p1

class AssertUtil {

  def waitForIt(terminated: => Boolean, progress: Int = 0, label: => String = "test"): Unit = {
    val limit = 5
    var n = 1
    var (dormancy, factor) = progress match {
      case 0 => (10000L, 5)
      case 1 => (250L, 4)
    }
    ()
  }
}


```

retronym commented:
I think I found a fix for the second crash, https://github.com/scala/scala/pull/9392/files#diff-0ade3e36bd7d74c3cf08289672961394062ee89d088703c4ba59e06a7be0136cR344 should have used `reverseMap`.

retronym committed:
Fixup partitionKeys / unused import

retronym committed:
Fix regression in box-unbox (in -opt:l:method)

Using 2.12.14-bin-8562bab or 2.13.x after this [merge](https://github.com/scala/scala/pull/9433)

```scala
// scalac: -opt:l:method
package p1
class AssertUtil {
  def waitForIt(terminated: => Boolean, progress: Int = 0, label: => String = "test"): Unit = {
    val limit = 5
    var n = 1
    var (dormancy, factor) = progress match {
      case 0 => (10000L, 5)
      case _ => (250L, 4)
    }
    ()
  }
}
```

```scala
// scalac: -opt:l:inline -opt-inline-from:<sources>
package p1; class C {
def f(b: java.lang.Byte) = {
  var box = 0
  @inline def writeBox: Unit = {
    box = 1
  }
  writeBox
}
}
```

Invalid bytecode generated, resulting in a later compiler crash.

The first incorrectly orders the compensating POP instructions inserted
for the `Long` and `Int` sized operands.

The second reproducer only fails when the patch in merged to 2.13.x, as
it relies on a particular bytecode shape that leaves a POP instruction
as the final consumer.

Discovered during https://github.com/scala/scala/pull/9495

Regression introduced in https://github.com/scala/scala/pull/9433.
This was merged _after_ 2.12.13 was released.

The enclosed test crafts the problematic bytecode for the second
case directly. The first is tested directly.

lrytz committed:
Merge pull request #9506 from retronym/topic/box-unbox-fix

Fix regression in box-unbox (in -opt:l:method)

dwijnand committed:
Add missing MiMa filters for NewRedBlackTree

dwijnand committed:
Merge pull request #9507 from regadas/12344_bt_mima_report

lrytz committed:
Merge pull request #9500 from retronym/cleanup-cleanup

Fix non-exhaustive match in Cleanup phase

retronym committed:
Merge commit '01a8c458ed' into merge-2.12.x-to-2.13.x-20210211

retronym committed:
Merge commit 'eab3ce8693' into merge-2.12.x-to-2.13.x-20210211

retronym committed:
Merge commit '2ee5f6c783' into merge-2.12.x-to-2.13.x-20210211

retronym committed:
Fixup box/unbox merge

retronym committed:
Fixup checkfile

retronym committed:
MiMa whitelist for RedBlackTree internal changes

dwijnand commented on project/MimaFilters.scala:
```diff
@@ -30,6 +30,13 @@ object MimaFilters extends AutoPlugin {
+    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.immutable.RedBlackTree#Tree.blackWithLeftRight"),
+    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.immutable.RedBlackTree.partitionKeys"),
+    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.immutable.RedBlackTree.partitionKeys"),
+    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.immutable.RedBlackTree#Tree.redWithLeft"),
```
These are already fixed in 2.12.x, btw (in a more general way): https://github.com/scala/scala/pull/9507/files, so no need to fast-back-track-port.

lrytz committed:
cleanups

lrytz reviewed (approved):
...whew 😰 thank you!

SethTisue commented:
marked "release-notes" because 2.12.14 isn't out yet, so the 2.13.5 notes may need to hit some highlights
